src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1482 @@
+
+use bitcoin_hashes::{Hash, hash160, sha256, sha256d, ripemd160};
+use miniscript::astelem::AstElem;
+use ::{Error, ToPublicKeyHash};
+use secp256k1::{self, Signature, VerifyOnly};
+use ::{ToPublicKey, Descriptor};
+use bitcoin;
+use ::Error::CouldnotEvaluate;
+
+/// Definition of Stack Element of the Witness stack used for interpretation of Miniscript.
+/// All stack elements with vec![] go to Dissatisfied and vec![1] are marked to Satisfied.
+/// Others are directly pushed as witness
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub enum StackElement<'stack> {
+    /// Result of a satisfied E/T/W/F
+    Satisfied,
+    /// Result of a dissatisfied E/W
+    Dissatisfied,
+    /// Input from the witness stack
+    Push(&'stack [u8]),
+}
+
+impl<'stack> StackElement<'stack>{
+
+    pub fn from(v: &'stack [u8]) -> StackElement<'stack> {
+        match v{
+            _ if *v == [1] => StackElement::Satisfied,
+            _ if *v == [] => StackElement::Dissatisfied,
+            _ => StackElement::Push(v)
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum HashLockType<'desc>{
+    ///SHA 256 hashlock
+    Sha256HashLock(&'desc sha256::Hash),
+    ///Hash 256 hashlock
+    Hash256HashLock(&'desc sha256d::Hash),
+    ///Hash160 hashlock
+    Hash160HashLock(&'desc hash160::Hash),
+    ///Ripemd160 hashlock
+    Ripemd160HashLock(&'desc ripemd160::Hash),
+}
+/// A satisfied Miniscript condition (Signature, Hashlock, Timelock)
+/// 'desc represents the lifetime of descriptor and `stack represents the lifetime of witness
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum SatisfiedConstraint<'desc, 'stack, Pk: 'desc, Pkh: 'desc> {
+    ///Public key and corresponding signature
+    PublicKey {
+        key: &'desc Pk,
+        sig: secp256k1::Signature,
+    },
+    ///Public keyhash, corresponding pubkey and signature
+    PublicKeyHasH {
```

Comment:
capitalization

Reply:
What should it be called otherwise?

Reply:
`PublicKeyHash`

Reply:
still unresolved

Reply:
I am really confused, this is `PublicKeyHash`.

Reply:
The last character in `PublicKeyHasH` should not be a capital

Reply:
oops, I thought really hard over this, still could not figure out until you told about this explicitly